crates/worker/src/partition/shuffle.rs

```diff
@@ -276,148 +272,174 @@ mod state_machine {
-                            );
+                        let (result, reader) = ready!(this.read_fut.poll_unpin(cx));
+                        *this.reader = Some(reader);
 
-                            *this.current_sequence_number = seq_number;
-
-                            let envelope =
-                                wrap_outbox_message_in_envelope(message, seq_number, this.metadata);
-                            let send_future = (this.send_operation)(envelope);
-
-                            this.state.set(State::Sending(send_future));
-                        } else {
-                            this.state.set(State::Idle);
+                        match result {
+                            Ok(None) => {
+                                this.state.set(State::Idle);
+                            }
+                            Ok(Some((sn, message))) => {
+                                let envelope =
+                                    wrap_outbox_message_in_envelope(message, sn, this.metadata);
+                                this.state.set(State::Ingesting {
+                                    ingest: this
+                                        .ingestion
+                                        .ingest(envelope.partition_key(), envelope),
+                                    sn,
+                                });
+                            }
+                            Err(err) => return Poll::Ready(err.into()),
                         }
                     }
-                    StateProj::Sending(send_future) => {
-                        send_future.await?;
+                }
+            }
+        }
+    }
 
-                        let successfully_shuffled_sequence_number = *this.current_sequence_number;
-                        *this.current_sequence_number += 1;
+    impl<T, R> Stream for ShuffleStream<T, R>
+    where
+        T: TransportConnect,
+        R: super::OutboxReader + Send + Sync + 'static,
+    {
+        type Item = Result<MessageIndex, anyhow::Error>;
 
-                        this.read_future.set(get_next_message(
-                            this.outbox_reader
-                                .take()
-                                .expect("outbox reader should be available"),
-                            *this.current_sequence_number,
-                        ));
-                        this.state.set(State::ReadingOutbox);
+        fn poll_next(mut self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
+            if let Poll::Ready(err) = self.as_mut().inner_poll_next(cx) {
+                return Poll::Ready(Some(Err(err)));
+            }
 
-                        return Ok(successfully_shuffled_sequence_number);
-                    }
+            let this = self.project();
+            let mut latest_committed_sn = None;
+            loop {
+                let Some(head) = this.inflight.front_mut() else {
```

Comment:
Can `this.inflight` potentially grow quite big if the head element cannot be delivered to the target partition but other messages (later ones) to other partitions will get delivered?